### PR TITLE
lua: vendored and sandboxed: v12

### PR DESCRIPTION
**Pull request by jasonish**
Lua, vendored and sandboxed.

The sandbox provides an allow list of functions. Some standard libraries aren't even loaded including:
- io
- os
- debug

An instruction count is put in a place. If a rule hits the instruction count a single log will be generated for that rule and a state incremented, much like when a script fails for other reasons.

Likewise for a blocked function.

However, functions to libraries that are not loaded don't have a special stat, as no stub is put in place for its functions, as the library is not even loaded.

Perhaps not quite ready for merge, but not in draft state either.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1855

### CODE REVIEW

File: doc/userguide/install.rst [hunk 1]
```diff
@@ -106,7 +106,7 @@ Recommended::
     sudo apt-get install autoconf automake build-essential ccache clang curl git \
                     gosu jq libbpf-dev libcap-ng0 libcap-ng-dev libelf-dev \
                     libevent-dev libgeoip-dev libhiredis-dev libjansson-dev \
-                    liblua5.1-dev libmagic-dev libnet1-dev libpcap-dev \
+                    liblua5.4-dev libmagic-dev libnet1-dev libpcap-dev \
```
Comment (victorjulien) [inline doc/userguide/install.rst:109]: is this still current?

File: doc/userguide/rules/lua-detection.rst [hunk 2]
```diff
@@ -15,7 +15,8 @@ Syntax:
 
 The script filename will be appended to your default rules location.
 
-The script has 2 parts, an init function and a match function. First, the init.
+The script has 2 parts, an init function and a match function. First, the init.
```
Comment (victorjulien) [inline doc/userguide/rules/lua-detection.rst:18]: this sentence makes less sense now: the "init" block is supposed to follow it directly I think

File: doc/userguide/rules/lua-detection.rst [hunk 3]
```diff
@@ -100,5 +101,28 @@ Entire script:
 
   return 0
 
+Sandbox and Available functions
+-------------------------------
+
+By default, the maximum memory and lua instruction count per execution of a detection rule will be limited.  Additionally,
+The following libraries and functions are blocked:
```
Comment (victorjulien) [inline doc/userguide/rules/lua-detection.rst:108]: should this now contain the allowlisted funcs instead?

File: src/util-lua-sandbox.c [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+ * See: https://www.lua.org/manual/5.4/manual.html#lua_Alloc
+ */
+static void *LuaAlloc(void *ud, void *ptr, size_t osize, size_t nsize)
+{
+    (void)ud;
+    (void)osize; /* not used */
+    SCLuaSbState *ctx = (SCLuaSbState *)ud;
+    if (nsize == 0) {
+        if (ptr != NULL) {
+            // ASSERT: alloc_bytes > osize
```
Comment (victorjulien) [inline src/util-lua-sandbox.c:61]: DEBUG_VALIDATE_BUG_ON? similar for other `ASSERT` comments?

File: src/util-lua-sandbox.c [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    } else {
+        // We can be a bit sloppy on the alloc limit since it's not supposed to be hit.
+        //  ASSERT: ctx->alloc_bytes + nsize > ctx->alloc_bytes
+        if (ctx->alloc_bytes + nsize > ctx->alloc_limit) {
+            // TODO: Trace in a better way
+            return NULL;
+        }
+        void *nptr = SCRealloc(ptr, nsize);
+        if (nptr != NULL) {
+            ctx->alloc_bytes += nsize;
```
Comment (victorjulien) [inline src/util-lua-sandbox.c:75]: looks suspicious for a few reasons:
- seems we need to add the delta of `osize` and `nsize`?
- can we shrink too?

File: rust/Cargo.toml.in [hunk 6]
```diff
@@ -69,6 +67,8 @@ time = "=0.3.20"
 
 suricata-derive = { path = "./derive", version = "@PACKAGE_VERSION@" }
 
+suricata-lua-sys = { git = "https://github.com/jasonish/suricata-lua-sys", version = "0.1.0-alpha.1" }
```
Comment (victorjulien) [inline rust/Cargo.toml.in:70]: think we can accept this for now, but we'd need a ticket to move this into the OISF github

File: .github/workflows/codeql.yml [hunk 7]
```diff
@@ -52,6 +52,7 @@ jobs:
        sudo apt-get install libssl-dev
        sudo apt-get install libpcre2-dev
        sudo apt-get install libjansson-dev
+       sudo apt-get install liblua5.4-dev
```
Comment (victorjulien) [inline .github/workflows/codeql.yml:55]: is this needed?

**Review by victorjulien — CHANGES_REQUESTED** [review]
Think we're close to merge, eager to actually :) Good stuff.

File: doc/userguide/rules/lua-detection.rst (hunk 3 above)
Comment (victorjulien) [reply doc/userguide/rules/lua-detection.rst:108]: wonder if we could utilize sphinx literalinclude to keep things in sync

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .uptime | 642 | 665 | 103.58% | 

Pipeline 20787

File: src/util-lua-sandbox.c (hunk 5 above)
Comment (jasonish) [reply src/util-lua-sandbox.c:75]: I haven't fully tested the memory allocation stuff.  I have an idea how it works, but still need to confirm, and will test this counter and add a stat and error message as well.

File: doc/userguide/rules/lua-detection.rst (hunk 3 above)
Comment (jasonish) [reply doc/userguide/rules/lua-detection.rst:108]: Probably should list the loaded libs and allowed functions, so yes, this needs cleanup. I don't want to directly pull from the C source code, I don't think it makes for nice documentation. In theory, it should only change when we do major upgrades to the Lua version as well.

**Comment by jasonish** [comment]
I'm not sure whats up with the utf8 failures in CI, my local environment is very close to one of the CI environments here and it doesn't quite make sense to me, unless Suricata it putting some invalid UTF-8 in the JSON application log. 

Converting to draft as I'm going to have to experiment by pushing a lot to CI via a pull request.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20799

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20800

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20801

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20802

**Comment by victorjulien** [comment]
Passes my QA, FYI

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20806

File: doc/userguide/install.rst (hunk 1 above)
Comment (jasonish) [reply doc/userguide/install.rst]: Nope. Fixed.

**Comment by jasonish** [comment]
> I'm not sure whats up with the utf8 failures in CI, my local environment is very close to one of the CI environments here and it doesn't quite make sense to me, unless Suricata it putting some invalid UTF-8 in the JSON application log.
> 
> Converting to draft as I'm going to have to experiment by pushing a lot to CI via a pull request.

Resolved.

File: src/util-lua-sandbox.c (hunk 5 above)
Comment (jasonish) [reply src/util-lua-sandbox.c:75]: Fixed up for next PR.

**Comment by jasonish** [comment]
Replaced by https://github.com/OISF/suricata/pull/11161.

**Pull request closed without merging**